### ISSUE TITLE
Enforcing event handlers as class fields with arrow functions

### ISSUE THREAD

**Issue by peteruithoven** [yannickcr/eslint-plugin-react#1799]
I'd like to enforce defining event handlers as class fields with arrow functions. 

Good:
``` js
class LoggingButton extends React.Component {
  handleClick = () => {
    console.log('this is:', this);
  }
  render() {
    return (
      <button onClick={this.handleClick}>Click me</button>
    );
  }
}
```

Bad:
``` js
class LoggingButton extends React.Component {
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }
  handleClick = () => {
    console.log('this is:', this);
  }
  render() {
    return (
      <button onClick={this.handleClick}>Click me</button>
    );
  }
}
```
Also bad:
``` js
class LoggingButton extends React.Component {
  handleClick() {
    console.log('clicked');
  }
  render() {
    return (
      <button onClick={this.handleClick}>Click me</button>
    );
  }
}
```

This is one of the workarounds mentioned in the [React Handling events documentation](https://reactjs.org/docs/handling-events.html):
> If you are using the experimental public class fields syntax, you can use class fields to correctly bind callbacks

Please let me know if there are existing rules that cover this. 

This rule is somewhat related to [react/jsx-no-bind](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-bind.md).

**Comment by ljharb** [comment]
This would be a terrible idea as class fields should never have arrow functions. The proper pattern is to constuructor-bind instance methods (your “bad” example)

**Comment by peteruithoven** [comment]
> class fields should never have arrow functions

Could you add some argumentation?

I did just learn that it impacts performance, Inheritance and Mockability:
https://medium.com/@charpeni/arrow-functions-in-class-properties-might-not-be-as-great-as-we-think-3b3551c440b1

**Comment by alexzherdev** [comment]
@ljharb genuinely curious, isn't constructor-binding instance methods achieving basically the same result as arrow functions on class fields? (which is, we still end up with a function attached to the instance as opposed to the prototype)

**Comment by ljharb** [comment]
@alexzherdev not exactly, because the "meat" of the function is still shared on the prototype, and thus optimized. The trivial bind-proxy can't be avoided, but at least that way we're minimizing the amount of un-shared code.

**Comment by Hypnosphi** [comment]
> the "meat" of the function is still shared on the prototype, and thus optimized

I heard that modern browsers optimize closures, so that basically the "meat" is stored once, but each instanse member gets a version with its own `this` from the closure, which is basically the same as binding in a constructor.

@ljharb do you know any good way to measure the difference in performance between bind in constructor vs instance member functions?

**Comment by ljharb** [comment]
@Hypnosphi it's entirely possible that browsers will be able to optimize away the performance consideration, but it's still bad to put functions in class fields both _conceptually_, and for testing, so performance really isn't the important point.

**Comment by damianobarbati** [comment]
+1 for this

**Comment by ljharb** [comment]
Class fields should forever not be class fields. If you are tempted to do that, do this instead:
```js
foo = this.foo.bind(this);
foo() {
  …
}
```
